models/multiphase/d3q27_pf_velocity/Dynamics.R

```diff
@@ -6,53 +6,76 @@ if (Options$q27){
-    AddField("gradPhiVal_x", stencil3d=2, group="gradPhi")
-    AddField("gradPhiVal_y", stencil3d=2, group="gradPhi")
-    AddField("gradPhiVal_z", stencil3d=2, group="gradPhi")
-    AddField("IsBoundary", stencil3d=1, group="debug_boundary")
+if (Options$staircaseimp) {
+    # actual normal directions
+    AddDensity(name="nw_actual_x", dx=0, dy=0, dz=0, group="nw_actual")
+    AddDensity(name="nw_actual_y", dx=0, dy=0, dz=0, group="nw_actual")
+    AddDensity(name="nw_actual_z", dx=0, dy=0, dz=0, group="nw_actual")
+
+    # Standard staircase improvement
+    AddDensity(name="coeff_v1", dx=0, dy=0, dz=0, group="st_interpolation")
+    AddDensity(name="coeff_v2", dx=0, dy=0, dz=0, group="st_interpolation")
+    AddDensity(name="coeff_v3", dx=0, dy=0, dz=0, group="st_interpolation")
+    AddDensity(name="triangle_index", dx=0, dy=0, dz=0, group="st_interpolation")
+
+    if (Options$tprec) {
+        # Staircase improvement with more precise second triangle
+        AddDensity(name="coeff2_v1", dx=0, dy=0, dz=0, group="st_interpolation")
+        AddDensity(name="coeff2_v2", dx=0, dy=0, dz=0, group="st_interpolation")
+        AddDensity(name="coeff2_v3", dx=0, dy=0, dz=0, group="st_interpolation")
+        AddDensity(name="triangle_index2", dx=0, dy=0, dz=0, group="st_interpolation")
+    }
+
+    # Make sure that those fields are now loaded
+    extra_save_iteration = c(extra_save_iteration, "nw_actual", "st_interpolation")
+    extra_load_iteration = c(extra_load_iteration, "nw_actual", "st_interpolation")
+    extra_load_phase = c(extra_load_phase, "nw_actual")
+    extra_fields_to_load_for_bc = c("nw_actual", "st_interpolation")
+}
 
-    AddDensity(name="TangentWallVector1_x", dx=0, dy=0, dz=0)
-    AddDensity(name="TangentWallVector2_x", dx=0, dy=0, dz=0)
-    AddDensity(name="TangentWallVector1_y", dx=0, dy=0, dz=0)
-    AddDensity(name="TangentWallVector2_y", dx=0, dy=0, dz=0)
-    AddDensity(name="TangentWallVector1_z", dx=0, dy=0, dz=0)
-    AddDensity(name="TangentWallVector2_z", dx=0, dy=0, dz=0)
-    AddField("PhaseF",stencil3d=3, group="PF")
 
-    AddStage("WallInit_CA"  , "Init_wallNorm", save=Fields$group %in% c("nw", "debug_boundary"))
-    AddStage("calcWall_CA"  , "calcWallPhase", save=Fields$name %in% c("PhaseF", "der_tangent_1_wall", "der_tangent_2_wall", "TangentWallVector1_x", "TangentWallVector2_x", "TangentWallVector1_y", "TangentWallVector2_y", "TangentWallVector1_z", "TangentWallVector2_z", "perpVal"), load=DensityAll$group %in% c("nw", "gradPhi", "PF"))
+AddDensity("IsSpecialBoundaryPoint", dx=0, dy=0, dz=0, group="solid_boundary")
+AddQuantity("SpecialBoundaryPoint", unit = 1)
 
-    AddStage('calcPhaseGrad', "calcPhaseGrad", load=DensityAll$group %in% c("g","h","Vel","nw", "PF"), save=Fields$group=="gradPhi")
-    AddStage('calcPhaseGrad_init', "calcPhaseGrad_init", load=DensityAll$group %in% c("g","h","Vel","nw", "PF"), save=Fields$group=="gradPhi")
+if (Options$altContactAngle){
```

Comment:
Could we change the option name from _altContactAngle to _geometric?

Reply:
Oh, yes, I completely forgot. The names became too familiar in my mind.